### ISSUE TITLE
Does not work with external Singular - path for dickson.lib

### ISSUE THREAD

**Issue by dimpase** [sagemath/p_group_cohomology#5]
Currently `dickson.lib` is installed by Sage's spkg-install to `$SAGE_LOCAL/...`.

An external installation of Singular, as worked on in https://trac.sagemath.org/ticket/29024, won't know about it.
A bulletproof way to deal with this is akin to what's done with `core.lib` in `src/sage/rings/function_field/function_field.py` - it is 
install by some Python magic in the "right place" (tm) and then its loading done via
```
            from sage.libs.singular.function import singular_function, lib
            from sage.env import SAGE_EXTCODE
            lib(SAGE_EXTCODE + '/singular/function_field/core.lib')
```

**Comment by dimpase** [comment]
@mkoeppe - could you comment on how exactly installation of `core.lib` is done?

**Comment by orlitzky** [comment]
Either `dickson.lib` needs to be moved into one of the sub-folders with a build system, or we will wind up hard-coding the path one way or another. That's the way it always was with `SINGULARPATH`, though, and while I say it like it's a bad thing it's the path of least resistance.

I.e. we add some init code to load `SAGE_EXTCODE + '/singular/p_group_cohomology/dickson.lib'`,  and then rely on the spkg-install file to `mkdir/cp` the file there.

**Comment by dimpase** [comment]
Why should it rely on SAGE_EXTCODE ? It is just a Python package installing a non-Python data, nothing unusual.

**Comment by orlitzky** [comment]
> Why should it rely on SAGE_EXTCODE ? 

It doesn't matter to me, I was only copying the existing example.

**Comment by mkoeppe** [comment]
Where is the actual code that loads `dickson.lib`?

**Comment by mkoeppe** [comment]
> Why should it rely on SAGE_EXTCODE ? It is just a Python package installing a non-Python data, nothing unusual.

The code for loading `dickson.lib` should of course not use the `SAGE_EXTCODE` variable. It should use `os.path.dirname(pGroupCohomology.__file__)` to determine where the package (and hence its package data) is installed.

**Comment by simon-king-jena** [comment]
Wouldn't `os.path.dirname(pGroupCohomology.__file__)` denote the location of the python module? Or worse, of the source file? That should be totally unrelated with the location in which `dickson.lib` is supposed to be installed. `dickson.lib` is Singular code and should be put where Singular finds it - it isn't relevant where the cython or python part of the module lives.

In other words, it would be best to ask Singular where its interpreted modules are supposed to be installed. I don't know how to make Singular answer that question, and that's why I inferred the correct location from `SAGE_EXTCODE`.

So, to be done:

- Remove the usage of `SAGE_EXTCODE` and ask Singular for the correct install location.
- Fix the fact that at least one of the Singular modules that the pGroupCohomology spkg provides is in a deprecated style (I don't know yet how Singular modules are supposed to be written nowadays).
- Cope with Singular's renaming `poly.lib -> polylib.lib`.

**Comment by simon-king-jena** [comment]
> ... that's why I inferred the correct location from `SAGE_EXTCODE`.

Or do I?? If I understand correctly, it was Dima's suggestion in the original post to use `SAGE_EXTCODE` similar to what's done in `src/sage/rings/function_field`. Anyway, I think the cohomology package should be made work with an external singular.

**Comment by mkoeppe** [comment]
> Wouldn't `os.path.dirname(pGroupCohomology.__file__)` denote the location of the python module? 

Yes, and this is exactly where I suggest that `dickson.lib` should be installed.
Then, when your Python code asks Singular to load this file, provide this location.
(Where does your code do this, by the way?)

**Comment by dimpase** [comment]
Singular can be told to load a library from any location specified by a
full path.

So you don't need to ask it where to install dickson.lib.

**Comment by mkoeppe** [comment]
>  it would be best to ask Singular where its interpreted modules are supposed to be installed. I don't know how to make Singular answer that question

The point is that there is no such location. When a system-installed Singular is use, if a user installs your package, the system location of the Singular library is not writable.

**Comment by simon-king-jena** [comment]
> > Wouldn't `os.path.dirname(pGroupCohomology.__file__)` denote the location of the python module?
> 
> Yes, and this is exactly where I suggest that `dickson.lib` should be installed.

No. It is a Singular module and should be where other Singular modules are, I believe.

> Then, when your Python code asks Singular to load this file, provide this location.
> (Where does your code do this, by the way?)

It does so in the standard way, i.e. by `singular.LIB("dickson")`, in `cohomology.pyx`. May better be moved to `auxiliary.py`, while we are at it.

**Comment by simon-king-jena** [comment]
> Where is the actual code that loads `dickson.lib`?

The code that INSTALLS `dickson.lib` is this:
`spkg-install.in:sdh_install singular_helper/dickson.lib "$SAGE_SHARE/singular/LIB/" || sdh_die "Error installing Singular helper module"`

The code that loads it is mentioned in my previous post.

**Comment by mkoeppe** [comment]
> It is a Singular module and should be where other Singular modules are, I believe.

Are users expected to use it directly from Singular? Or is it only intended to be invoked from your Python package?

**Comment by dimpase** [comment]
Other Singular modules might be in a location the user installing your
package cannot even write to. So this is not feasible.

 Singular provides methods for not specifying full path to the location,
but you don't need any of them.

**Comment by simon-king-jena** [comment]
> The point is that there is no such location.

Well, if you do the equivalent of `ls ~/Sage/git/sage/local/share/singular/LIB/` in your machine, you may find there *is* a standard location.

> When a system-installed Singular is use, if a user installs your package, the system location of the Singular library is not writable.

OK, that's a valid point.

**Comment by mkoeppe** [comment]
> > The point is that there is no such location.
> 
> Well, if you do the equivalent of `ls ~/Sage/git/sage/local/share/singular/LIB/` in your machine, you may find there _is_ a standard location.

It is the location that Sage-the-distribution uses because it controls where it installs its copy of Singular.

**Comment by simon-king-jena** [comment]
> Are users expected to use it directly from Singular? Or is it only intended to be invoked from your Python package?

I wouldn't exclude it. If someone is interested in Dickson invariants, the code there may be useful. But in that case, admittedly one could ask the Singular team if the module couldn't be included upstream.

**Comment by simon-king-jena** [comment]
> It is the location that Sage-the-distribution uses because it controls where it installs its copy of Singular.

Exactly. And an external installation of Singular was able to answer where the corresponding standard location can be found, then the only problem for an installation would be write permission.

Write permission could be a hard problem, though. That's why I wouldn't mind to move 'dickson.lib` to a location that is controlled by the package.

**Comment by dimpase** [comment]
Just use the Python directory of the package - and use `LIB` with full path, e.g.
```
$ Singular 
                     SINGULAR                                 /
 A Computer Algebra System for Polynomial Computations       /   version 4.2.1
                                                           0<
 by: W. Decker, G.-M. Greuel, G. Pfister, H. Schoenemann     \   May 2021
FB Mathematik der Universitaet, D-67653 Kaiserslautern        \
> LIB "/mnt/opt/Sage/sage-dev/local/share/singular/LIB/dickson.lib";
// ** loaded /mnt/opt/Sage/sage-dev/local/share/singular/LIB/dickson.lib 
// ** library /mnt/opt/Sage/sage-dev/local/share/singular/LIB/dickson.lib has old format. This format is still accepted,
// ** but for functionality you may wish to change to the new
// ** format. Please refer to the manual for further information.
// ** loaded /usr/bin/../share/singular/LIB/polylib.lib (4.2.0.0,Dec_2020)
// ** redefining ringweights (LIB "/mnt/opt/Sage/sage-dev/local/share/singular/LIB/dickson.lib";)
// ** redefining ringweights (LIB "/mnt/opt/Sage/sage-dev/local/share/singular/LIB/dickson.lib";)
// ** loaded /usr/bin/../share/singular/LIB/ring.lib (4.1.2.0,Feb_2019)
...
```

**Comment by mkoeppe** [comment]
> in that case, admittedly one could ask the Singular team if the module couldn't be included upstream.

Yes, my impression is that this is the way that Singular packages are distributed -- they become part of the Singular distribution.

**Comment by simon-king-jena** [comment]
> Yes, my impression is that this is the way that Singular packages are distributed -- they become part of the Singular distribution.

On the other hand, after having a closer look at it, suspiciously often I mention p_group_cohomology in `dickson.lib`'s comments: I compute Dickson invariants that are modified in a way that is used in the cohomology package, and also there are functions to find elements in the cohomology of a group that simultaneously restrict to elements in the cohomology of an elementary abelian subgroup. So, only a small part of the module would be useful for people that work with Dickson invariants.

**Comment by mkoeppe** [comment]
In any case, contributing some version or part of it to Singular would be a longer-term effort; so it's probably best to focus on the approach of shipping `dickson.lib` as `package_data` of your Python package, which is immediately actionable

**Comment by simon-king-jena** [comment]
> In any case, contributing some version or part of it to Singular would be a longer-term effort; so it's probably best to focus on the approach of shipping `dickson.lib` as `package_data` of your Python package, which is immediately actionable

Do I understand correctly: `package_data` would be in `setup.py`, not in `spkg-install.in` and not in `makefile.am`, right?

**Comment by mkoeppe** [comment]
> Do I understand correctly: `package_data` would be in `setup.py`

Yes

**Comment by mkoeppe** [comment]
> @mkoeppe - could you comment on how exactly installation of `core.lib` is done?

Via `package_data` in https://github.com/sagemath/sage/blob/develop/pkgs/sagemath-standard/setup.py#L101

**Comment by dimpase** [comment]
> > @mkoeppe - could you comment on how exactly installation of `core.lib` is done?
> 
> Via `package_data` in https://github.com/sagemath/sage/blob/develop/pkgs/sagemath-standard/setup.py#L101

I don't see any `core.lib` there.

**Comment by mkoeppe** [comment]
It's now done in `src/setup.cfg.m4` via `ext_data/singular/function_field/*`

**Comment by dimpase** [comment]
I think that `core.lib` get installed via `setup.cfg` magic:
```
...
[options.package_data]

...

sage =
    ext_data/*
    ext_data/kenzo/*
    ext_data/singular/*
    ext_data/singular/function_field/*
...
```

**Comment by mkoeppe** [comment]
Yes, and `gap_helper`, `singular_helper` should likewise be moved to https://github.com/sagemath/p_group_cohomology/tree/master/src/pGroupCohomology
and declared as `package_data`.

**Comment by dimpase** [comment]
fixed in v3.3.3
